src/server/editorServices.ts

```diff
@@ -855,65 +858,247 @@ namespace ts.server {
+                this.mapOfConfigFilePresence.set(project.canonicalConfigFilePath, {
+                    exists: true,
+                    trackingOpenFilesMap: createMap<boolean>()
+                });
+            }
+        }
+
+        private configFileExistenceTracksInferredRoot(configFilePresenceInfo: ConfigFileExistence) {
+            return forEachEntry(configFilePresenceInfo.trackingOpenFilesMap, (value, __key) => value);
+        }
+
+        private setConfigFilePresenceByClosedConfigFile(closedProject: ConfiguredProject) {
+            const configFilePresenceInfo = this.mapOfConfigFilePresence.get(closedProject.canonicalConfigFilePath);
+            Debug.assert(!!configFilePresenceInfo);
+            const trackingOpenFilesMap = configFilePresenceInfo.trackingOpenFilesMap;
+            if (trackingOpenFilesMap.size) {
+                const configFileName = closedProject.getConfigFilePath();
+                if (this.configFileExistenceTracksInferredRoot(configFilePresenceInfo)) {
+                    Debug.assert(!configFilePresenceInfo.configFileWatcher);
+                    configFilePresenceInfo.configFileWatcher = this.addFileWatcher(
+                        WatchType.ConfigFileForInferredRoot, /*project*/ undefined, configFileName,
+                        (_filename, eventKind) => this.onConfigFileChangeForOpenScriptInfo(configFileName, eventKind)
+                    );
+                    this.logConfigFileWatchUpdate(configFileName, configFilePresenceInfo, ConfigFileWatcherStatus.UpdatedCallback);
+                }
+            }
+            else {
+                // There is no one tracking anymore. Remove the status
+                this.mapOfConfigFilePresence.delete(closedProject.canonicalConfigFilePath);
+            }
+        }
+
+        private logConfigFileWatchUpdate(configFileName: NormalizedPath, configFilePresenceInfo: ConfigFileExistence, status: ConfigFileWatcherStatus) {
+            if (this.logger.loggingEnabled()) {
+                const inferredRoots: string[] = [];
+                const otherFiles: string[] = [];
+                configFilePresenceInfo.trackingOpenFilesMap.forEach((value, key: Path) => {
+                    const info = this.getScriptInfoForPath(key);
+                    if (value) {
+                        inferredRoots.push(info.fileName);
+                    }
+                    else {
+                        otherFiles.push(info.fileName);
+                    }
+                });
+                const watchType = status === ConfigFileWatcherStatus.UpdatedCallback ||
+                    status === ConfigFileWatcherStatus.ReloadingFiles ||
+                    status === ConfigFileWatcherStatus.ReloadingInferredRootFiles ?
+                    (configFilePresenceInfo.configFileWatcher ? WatchType.ConfigFileForInferredRoot : WatchType.ConfigFilePath) :
+                    "";

```

Comment:
This will look strange when included in the log, as in `ConfigFilePresence Config file for the inferred project root::  ...`